aioxmpp/hashes.py

```diff
@@ -256,8 +311,92 @@ class HashesParent(xso.XSO):
     )
 
 
+class HashUsedType(xso.AbstractElementType):
+    @classmethod
+    def get_xso_types(cls):
+        return [HashUsed]
+
+    def unpack(self, obj):
+        return obj.algo
+
+    def pack(self, item):
+        return HashUsed(item)
+
+
+class HashesUsedParent(xso.XSO):
+    """
+    Mix-in class for XSOs which use :class:`HashUsed` children.
+
+    .. attribute:: algos
+
+        A list of hash algorithms.
+    """
+
+    algos = xso.ChildValueList(
+        type_=HashUsedType(),
+    )
+
+
 default_hash_algorithms = {
     algo
     for algo in ["sha-256", "sha3-256", "blake2b-256"]
     if is_algo_supported(algo)
 }
+
+
+class HashService(service.Service):
+    """
+    The service component of the :xep:`300` support. This service registeres
+    the features and allows to query the hash functions supported by us and
+    a remote entity:
+
+    .. automethod:: select_common_hashes
+    """
+    ORDER_AFTER = [
+        disco.DiscoClient,
+        disco.DiscoServer,
+    ]
+
+    hashes_feature = disco.register_feature(namespaces.xep0300_hashes2)
+
+    def __init__(self, client, **kwargs):
+        super().__init__(client, **kwargs)
+        self._disco_client = self.dependencies[disco.DiscoClient]
+        self._disco_server = self.dependencies[disco.DiscoServer]
+
+        for feature in SUPPORTED_HASH_FEATURES:
+            self._disco_server.register_feature(feature)
+
+    @asyncio.coroutine
+    def _shutdown(self):
+        for feature in SUPPORTED_HASH_FEATURES:
+            self._disco_server.unregister_feature(feature)
+        yield from super()._shutdown()
+
+    @asyncio.coroutine
+    def select_common_hashes(self, other_entity):
```

Comment:
I’m not sure if we might want to have this as a function instead (which takes a DiscoClient and a target JID as arguments). This would alleviate the service of the DiscoServer dependency.

Thoughts?

Reply:
As a coroutine function in module scope? That sounds like a good idea (although it does make the HashService even more degenerate).

Reply:
Modifying state of a dependent service is a good use for a Service class.

Reply:
There is an argument against moving the coroutine out of the service class: consistency. If the HashService is not summoned and we determine the common hashes by service discovery, then we will see the correct list of common hashes, while the other would determine no commonly supported hashes – so we should only be able to use this functionality if we also publish our own hash-features.

Reply:
Ah, that probably makes sense.